### PR TITLE
fix Issue 23058 - importC: cannot take address inside multi-dimension…

### PR DESCRIPTION
**Pull request by WalterBright**
…al array at compile time

### CODE REVIEW

**Comment by WalterBright** [comment]
Test suite failures are networking problems with Azure.

File: src/dmd/optimize.d [hunk 1]
```diff
@@ -558,6 +558,41 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
...
+                 */
+                sinteger_t index = ae.e2.toInteger();
+                auto ae1 = ae.e1.isIndexExp();          // ae1 is a[i]
+                if (auto ts = ae1.type.isTypeSArray())
+                {
+                    sinteger_t dim = ts.dim.toInteger();
+
+                    if (index < 0 || index > dim)
+                    {
+                        e.error("array index %lld is out of bounds `[0..%lld]`", index, dim);
```
Comment (dkorpel) [inline src/dmd/optimize.d:573]: Not covered by tests

File: src/dmd/optimize.d [hunk 2]
```diff
@@ -558,6 +558,41 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
...
+                        return error();
+                    }
+
+                    import core.checkedint : mulu;
+                    bool overflow;
+                    const offset = mulu(index, ts.nextOf().size(e.loc), overflow); // offset = index*size
+                    if (overflow)
+                    {
+                        e.error("array offset overflow");
```
Comment (dkorpel) [inline src/dmd/optimize.d:582]: Not covered by tests

File: test/fail_compilation/fail13902.d [hunk 3]
```diff
@@ -57,7 +57,7 @@ fail_compilation/fail13902.d(88): Error: Using the result of a comma expression
 fail_compilation/fail13902.d(75): Error: returning `& x` escapes a reference to parameter `x`
 fail_compilation/fail13902.d(76): Error: returning `&s1.v` escapes a reference to parameter `s1`
 fail_compilation/fail13902.d(81): Error: returning `& sa1` escapes a reference to parameter `sa1`
-fail_compilation/fail13902.d(82): Error: returning `&sa2[0][0]` escapes a reference to parameter `sa2`
```
Comment (dkorpel) [inline test/fail_compilation/fail13902.d:60]: This PR introduces a regression in D error messages.
```D
int* f()
{
    int[2][2] sa2;
    return &sa2[1][1];
}
```

> returning `(& sa2+12)` escapes a reference to local variable `sa2`

That's not what's being returned.

File: src/dmd/optimize.d [hunk 4]
```diff
@@ -558,6 +558,41 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
         }
         else if (auto ae = e.e1.isIndexExp())
         {
+            if (ae.e2.isIntegerExp() && ae.e1.isIndexExp())
+            {
+                /* Rewrite `(a[i])[index]` to `(&a[i]) + index*size`
```
Comment (dkorpel) [inline src/dmd/optimize.d:563]: The comment should say why this is done

**Review by dkorpel — CHANGES_REQUESTED** [review]

**Comment by RazvanN7** [comment]
ping @WalterBright

File: test/fail_compilation/fail13902.d (hunk 3 above)
Comment (WalterBright) [reply test/fail_compilation/fail13902.d:60]: yes, it is what is being returned.

File: src/dmd/optimize.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/optimize.d:563]: The reason is the same as for every other rewrite - to put it in simpler terms that the rest of the code can handle. I.e. "lowering".

File: src/dmd/optimize.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/optimize.d:582]: It's difficult to come up with a test for that.

Comment (WalterBright) [reply src/dmd/optimize.d:582]: In many cases, I have been stymied by trying to trip these, as the overflows happen before it gets here. But the check should still be there, for defensive programming purposes.

**Comment by WalterBright** [comment]
gawd, not again.

**Comment by WalterBright** [comment]
rebasing it back to stable made an even worse mess.

**Comment by WalterBright** [comment]
Rebooted as https://github.com/dlang/dmd/pull/14094
